### ISSUE TITLE
Completely deleting a downloadable product renders the Downloadable Product Permissions' "Grant Access" unusable

### ISSUE THREAD

**Issue by billrobbins** [woocommerce/woocommerce#30300]
**Prerequisites (mark completed items with an [x]):**
- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
An order is placed for a downloadable product.  That product is later deleted and removed from the trash.  At that point, it becomes impossible to use "grant permission" to add downloadable permissions for other products added to the order.  

**Expected behavior**
I expect to be able to "grant permission" for any product that has been added to an order.

**Actual behavior**
Once I click the grant permission button, there is a 500 error on the AJAX call.  It returns this error:

```
2021-07-16T16:36:27+00:00 CRITICAL Uncaught Error: Call to a member function get_id() on bool in /Users/bill/Local Sites/quick-dev/app/public/wp-content/plugins/woocommerce/includes/class-wc-ajax.php:802
Stack trace:
#0 /Users/bill/Local Sites/quick-dev/app/public/wp-includes/class-wp-hook.php(292): WC_AJAX::grant_access_to_download('')
#1 /Users/bill/Local Sites/quick-dev/app/public/wp-includes/class-wp-hook.php(316): WP_Hook->apply_filters('', Array)
#2 /Users/bill/Local Sites/quick-dev/app/public/wp-includes/plugin.php(484): WP_Hook->do_action(Array)
#3 /Users/bill/Local Sites/quick-dev/app/public/wp-admin/admin-ajax.php(187): do_action('wp_ajax_woocomm...')
#4 {main}
  thrown in /Users/bill/Local Sites/quick-dev/app/public/wp-content/plugins/woocommerce/includes/class-wc-ajax.php on line 802
```

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Create two downloadable products.
2. Add one to the cart and checkout with it.  Set the order to processing or completed to add download permissions.
3. Delete the product from the store (not the order) and empty it from the trash.
4. Go back to the order and change it to on hold.
5. Add the other downloadable product to the order and update it.
6. Now try to add download permissions for the new product using the "Downloadable product permissions" box.
7. Notice how this fails.

**Screenshots**
![download-on-order](https://user-images.githubusercontent.com/1138631/125983082-dcde7096-3458-46b2-834f-2f2c7fe73385.png)

![after-deleted](https://user-images.githubusercontent.com/1138631/125983097-d7357ff5-3975-4462-90b6-5a2f2e9c17b8.png)

![adding-new-item](https://user-images.githubusercontent.com/1138631/125983105-e93c175b-af96-4dca-af04-fdde048d4839.png)

**Workaround**
Using "regenerate download permissions" from the order actions menu works.  It will add download permission for the new product that was added to the order.

**Isolating the problem (mark completed items with an [x]):**
- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**WordPress Environment**

<details>
```
### WordPress Environment ###

WordPress address (URL): http://quick-dev.local
Site address (URL): http://quick-dev.local
WC Version: 5.5.1
REST API Version: ✔ 5.5.1
WC Blocks Version: ✔ 5.3.2
Action Scheduler Version: ✔ 3.2.1
WC Admin Version: ✔ 2.4.1
Log Directory Writable: ✔
WP Version: 5.7.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.0.0
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 7.54.0
LibreSSL/2.6.5

SUHOSIN Installed: –
MySQL Version: 5.7.28
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 5.5.1
WC Database Prefix: wp_
Total Database Size: 4.57MB
Database Data Size: 3.08MB
Database Index Size: 1.49MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.05MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.14MB + Index: 0.09MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 27
page: 7
post: 2
product: 19
product_variation: 7
revision: 2
shop_order: 4

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (1) ###

WooCommerce: by Automattic – 5.5.1

### Inactive Plugins (2) ###

String Locator: by Clorith – 2.4.2
Stripe Customer Information Remover: by Bill Robbins – 1.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 3.7.0
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 19
Oldest: 2021-07-16 16:21:20 +0000
Newest: 2021-07-16 16:37:52 +0000


### Status report information ###

Generated at: 2021-07-16 16:59:24 +00:00

```
</details>

**Comment by rodelgc** [comment]
Hi @billrobbins,

Thank you for opening the issue! It requires further feedback from the WooCommerce Core team. I am adding the `needs developer feedback` label to this issue so that the Core team could take a look. 

Please note it may take a few days for them to get to this issue. Thank you for your patience.

**Comment by barryhughes** [comment]
Hi @billrobbins,

Thanks for reporting this and for sharing the steps to replicate.

We can't actually see the same thing, and are curious if you still see this when using our latest release (or else the latest `trunk` code)? It's just possible this specific issue was resolved via #29884, so if you have a moment to confirm that would be appreciated :-)

**Comment by billrobbins** [comment]
Hey @barryhughes 

I tested just now with 5.6.1 - Beta 1 and I can't reproduce this anymore.  Looks like this is now fixed :)

**Comment by barryhughes** [comment]
Great—thank you for testing+confirming :+1:
